iceoryx_hoofs/include/iceoryx_hoofs/error_handling/error_handler.hpp

```diff
@@ -0,0 +1,145 @@
+constexpr uint16_t USER_DEFINED_MODULE_IDENTIFIER{256};
+
+constexpr uint8_t ERROR_ENUM_OFFSET_IN_BITS{16};
+
+#define CREATE_ICEORYX_ERROR_ENUM(name) name,
+#define CREATE_ICEORYX_ERROR_STRING(name) #name,
+
+/// @brief the available error levels
+/// FATAL
+/// - Log message with FATAL
+/// - Application cannot recover from that error and is terminated
+/// - Assert (in DEBUG) and terminate
+/// - Reporting code must handle this and continue or go to a save state. Error handler could return (e.g. in test)
+/// SEVERE
+/// - Log message with ERROR
+/// - Application can still run. Error is reported
+/// - Assert in DEBUG, in RELEASE continue to run
+/// - Reporting code must handle this and continue
+/// MODERATE
+/// - Log message with ERROR
+/// - Application can still run. Error is reported
+/// - NO assert
+/// - Reporting code must handle this and continue
+enum class ErrorLevel : uint32_t
+{
+    /// Log error entry + Assert + terminate
+    FATAL,
+    /// warning log entry + Assert
+    SEVERE,
+    /// warning log entry
+    MODERATE
+};
+
+/// @brief Howto use the error handler correctly
+///     1.) Use the macro ICEORYX_ERRORS(error) to create the enum for your component and
+///             add new errors like:
+///             error(MODULE_NAME__MY_FUNKY_ERROR)
+///         Attention: Create an error after the following convention:
+///             MODULE_NAME__A_CLEAR_BUT_SHORT_ERROR_DESCRIPTION
+///         And a long name is alright!
+///
+///     2.) Add a new FOO_MODULE_IDENTIFIER at the top of this file or use
+///         USER_DEFINED_MODULE_IDENTIFIER as the first entry in the enum:
+///         - NO_ERROR = FOO_MODULE_IDENTIFIER << ERROR_ENUM_OFFSET_IN_BITS
+///         - NO_ERROR = USER_DEFINED_MODULE_IDENTIFIER << ERROR_ENUM_OFFSET_IN_BITS
+///
+///     3.) Specialize the following methods for your NewEnumErrorType:
+///         - const char* asStringLiteral(const NewEnumErrorType error)
+///
+///     4.) Call errorHandler(Error::MODULE_NAME__MY_FUNKY_ERROR);
+///         The defaults for ErrorLevel can also be overwritten:
+///             errorHandler(
+///                 Error::MODULE_NAME__MY_FUNKY_ERROR,
+///                 ErrorLevel::MODERATE
+///             );
+///
+/// @code
+/// // definition of error_handling.hpp for module 'foo'
+/// #define FOO_ERRORS(error) '\'
+///     error(PRETTY_CLASS__DIVISION_BY_ZERO)
+///
+/// enum class FooError : uint32_t
+/// {
+///     NO_ERROR = FOO_MODULE_IDENTIFIER << ERROR_ENUM_OFFSET_IN_BITS,
+///     FOO_ERRORS(CREATE_ICEORYX_ERROR_ENUM)
+/// };
+///
+/// const char* asStringLiteral(const FooError error) noexcept;
+///
+/// // usage of the module
+/// #include "module_foo/error_handling.hpp"
+///
+/// class PrettyClass {
+///     float division(float a, float b) {
+///         if ( b == 0.0f ) {
+///             errorHandler(FooError::PRETTY_CLASS__DIVISION_BY_ZERO);
+///         }
+///     }
```

Comment:
Well, we provide a middleware and not a computation library ;)
But yes, it might be a good idea to have correct code in the example. Maybe also with integer instead of floats. I'm not sure if there are implementations where `+0.0 != -0.0`

Reply:
This is something for a follow-up PR.